stor/tests/test_integration_dx.py

```diff
@@ -0,0 +1,320 @@
+
+            # Create files filled with random data.
+            path1 = tmp_d / 'dir1'
+            os.mkdir(path1)
+            self.create_dataset(path1, num_test_objs, 10)
+
+            # Create empty dir and file.
+            path2 = tmp_d / 'dir2'
+            os.mkdir(path2)
+            os.mkdir(path2 / 'my_dir')
+            open(path2 / 'my_dir' / 'included_file', 'w').close()
+            open(path2 / 'my_dir' / 'excluded_file', 'w').close()
+            os.mkdir(path2 / 'my_dir' / 'included_dir')
+            os.mkdir(path2 / 'my_dir' / 'excluded_dir')
+
+            # Create file in the top level directory.
+            open(tmp_d / 'top_level_file', 'w').close()
+
+            to_upload = [
+                'dir1',
+                'dir2/my_dir/included_file',
+                'dir2/my_dir/included_dir',
+                'top_level_file',
+            ]
+            with tmp_d:
+                dx_p = self.test_dir / 'subdir'
+                dx_p.upload(to_upload)
+
+            # Validate the contents of the manifest file
+            uploaded_contents = stor.list(dx_p)
+            expected_contents = [
+                Path('dir1') / name
+                for name in self.get_dataset_obj_names(num_test_objs)
+            ]
+            expected_contents.extend([
+                'dir2/my_dir/included_file',
+                'top_level_file',
+            ])
+
+            expected_contents = [dx_p / c for c in expected_contents]
+            self.assertEquals(set(uploaded_contents), set(expected_contents))
+
+            empty_dir = dx_p / 'dir2/my_dir/included_dir'
+            self.assertTrue(stor.isdir(empty_dir))
+
+    def test_upload_download_remove(self):
+        num_test_objs = 10
+        min_obj_size = 50
+        with NamedTemporaryDirectory(change_dir=True) as tmp_d:
+            self.create_dataset(tmp_d, num_test_objs, min_obj_size)
+            self.test_dir.upload(['.'])
+
+        which_obj = self.get_dataset_obj_names(num_test_objs)[-1]
+        dx_p = self.test_dir / which_obj
+        file_h = dxpy.DXFile(dxid=dx_p.canonical_resource,
+                             project=dx_p.canonical_project)
+        file_h.wait_on_close(20)  # wait for file to go to closed state
+
+        for which_obj in self.get_dataset_obj_names(num_test_objs):
+            self.assertTrue((self.test_dir / which_obj).exists())
+
+        with NamedTemporaryDirectory(change_dir=True) as tmp_d:
+            self.test_dir.download(tmp_d)
+            for which_obj in self.get_dataset_obj_names(num_test_objs):
+                self.assertCorrectObjectContents(which_obj, which_obj, min_obj_size)
+                (self.test_dir / which_obj).remove()
+
+                # consistency check
+                while (self.test_dir / which_obj).exists():
+                    time.sleep(.5)
+                self.assertFalse((self.test_dir / which_obj).exists())
+
+    @skipIf(six.PY3, 'dxpy3 assumes utf-8 encoding, not suitable for gzip')
```

Comment:
it shouldn't assume utf-8 encoding if you open with `rb` - can you put up a small example on the GH dx-toolkit bug tracker to demonstrate issue? (e.g., something like creating and uploading gzipped file that you then try to open using `mode='rb'`)

pretty clear (but small) bug

Reply:
There is no mode parameter in `dxpy.DXFile.read()`. For stor, if mode is 'r', `DXPath.read_object()` is called, and if mode is 'rb', `DXPath.read_object().decode()` is invoked. `DXPath.read_object()` and hence `dxpy.DXFile.read()` is supposed to return the raw bytes. They return the raw bytes by using `decode('utf'-8')` (it's hardcoded).

I opened the issue at https://github.com/dnanexus/dx-toolkit/issues/426 . Please take a look.